File: common/src/main/java/dev/obscuria/fragmentum/v2/api/common/signal/Signal.java
package dev.obscuria.fragmentum.v2.api.common.signal;

import org.jetbrains.annotations.Nullable;

@SuppressWarnings("unused")
public interface Signal<T> {

    default void connect(T listener) {
        connect(null, false, null, listener);
    }

    default void connect(@Nullable Object source, T listener) {
        connect(source, false, null, listener);
    }

    default void connect(boolean oneShot, T listener) {
        connect(null, false, null, listener);
    }

    default void connect(@Nullable Object source, boolean oneShot, T listener) {
        connect(source, oneShot, null, listener);
    }

    void connect(@Nullable Object source, boolean oneShot, @Nullable Signal0 breaker, T listener);

    void disconnect(Object source);
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType")

package dev.obscuria.fragmentum.tools.signal

abstract class Signal<T> {

    fun connect(listener: T): Unit = error("API stub")

    fun connect(key: Any?, listener: T): Unit = error("API stub")

    fun connect(oneShot: Boolean, listener: T): Unit = error("API stub")

    fun connect(key: Any?, oneShot: Boolean, listener: T): Unit = error("API stub")

    fun connect(listener: T, key: Any? = null, oneShot: Boolean = false, breaker: Signal0? = null): Unit = error("API stub")

    infix fun disconnect(key: Any): Unit = error("API stub")
}